x-pack/plugin/core/src/internalClusterTest/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDeciderIT.java

```diff
@@ -67,6 +95,195 @@ public void testDefaultIndexAllocateToContent() {
+            .setSettings(Settings.builder().put("index.routing.allocation.exclude._name", coldNodeName).build())
+            .get();
+
+        assertBusy(() -> assertPrimaryShardIsAllocatedInNodeWithRole(0, DiscoveryNodeRole.DATA_COLD_NODE_ROLE));
+
+        startColdOnlyNode(newColdDesiredNode.externalId());
+
+        ensureGreen(index);
+
+        assertBusy(() -> assertPrimaryShardIsAllocatedInNode(0, newColdDesiredNode));
+
+        internalCluster().stopNode(coldNodeName);
+
+        ensureGreen(index);
+    }
+
+    public void testDesiredNodesAreTakenIntoAccountInAutoExpandReplicas() throws Exception {
+        final var masterDesiredNode = desiredNode(internalCluster().getMasterName(), DiscoveryNodeRole.MASTER_ROLE);
+        final int numberOfColdNodes = randomIntBetween(2, 5);
+        final List<DesiredNode> coldDesiredNodes = new ArrayList<>();
+        for (int i = 0; i < numberOfColdNodes; i++) {
+            final var coldDesiredNode = desiredNode("cold-node-" + i, DiscoveryNodeRole.DATA_COLD_NODE_ROLE);
+            coldDesiredNodes.add(coldDesiredNode);
+            startColdOnlyNode(coldDesiredNode.externalId());
+        }
+        final int numberOfWarmNodes = randomIntBetween(numberOfColdNodes + 1, 10);
+        final List<DesiredNode> warmDesiredNodes = new ArrayList<>();
+        for (int i = 0; i < numberOfWarmNodes; i++) {
+            final var warmDesiredNode = desiredNode("warm-node-" + i, DiscoveryNodeRole.DATA_WARM_NODE_ROLE);
+            warmDesiredNodes.add(warmDesiredNode);
+            startWarmOnlyNode(warmDesiredNode.externalId());
+        }
+        final List<DesiredNode> desiredNodesWithWarmAndColdTier = new ArrayList<>();
+        desiredNodesWithWarmAndColdTier.addAll(warmDesiredNodes);
+        desiredNodesWithWarmAndColdTier.addAll(coldDesiredNodes);
+        desiredNodesWithWarmAndColdTier.add(masterDesiredNode);
+
+        updateDesiredNodes(desiredNodesWithWarmAndColdTier);
+
+        client().admin()
+            .indices()
+            .prepareCreate(index)
+            .setWaitForActiveShards(0)
+            .setSettings(
+                Settings.builder()
+                    .put(DataTier.TIER_PREFERENCE, String.join(",", DataTier.DATA_COLD, DataTier.DATA_WARM))
+                    .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 0)
+                    .put(INDEX_AUTO_EXPAND_REPLICAS_SETTING.getKey(), "0-all")
+            )
+            .get();
+
+        var replicas = client().admin()
+            .indices()
+            .prepareGetIndex()
+            .setIndices(index)
+            .get()
+            .getSetting(index, INDEX_NUMBER_OF_REPLICAS_SETTING.getKey());
+
+        assertThat(Integer.parseInt(replicas), is(equalTo(numberOfColdNodes - 1)));
+
+        final List<DesiredNode> desiredNodesWithoutColdTier = new ArrayList<>(warmDesiredNodes);
+        desiredNodesWithoutColdTier.add(masterDesiredNode);
+
+        updateDesiredNodes(desiredNodesWithoutColdTier);
+
```

Comment:
We have a small race here, since we execute the reroute asynchronously after updating the desired nodes, it's possible to break the invariants in:
```
    private void reroute(RoutingAllocation allocation) {
        assert hasDeadNodes(allocation) == false : "dead nodes should be explicitly cleaned up. See disassociateDeadNodes";
        assert AutoExpandReplicas.getAutoExpandReplicaChanges(allocation.metadata(), () -> allocation).isEmpty()
            : "auto-expand replicas out of sync with number of nodes in the cluster";
```
i.e: if we execute a reroute after updating the desired nodes, the test fail.

I haven't checked if we have the same issues just after adding a new node, but I think it might be possible?

Reply:
> I am not entirely sure if you mean that the two invariants mentioned can break or whether you are arguing why the subsequent assertion is in an assertBusy, can you clarify?

I meant that the invariant `assert AutoExpandReplicas.getAutoExpandReplicaChanges(allocation.metadata(), () -> allocation).isEmpty()` can break if there's a reroute executed before the reroute that we trigger asynchronously after updating the desired nodes.